File: backend/app/models/monitoring.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.config.settings import MonitoringAlertSettings
from app.models.observability import ObservabilityHealth


class MonitoringLLMOverview(BaseModel):
    logical_call_count: int = 0
    provider_request_count: int = 0
    retry_request_count: int = 0
    failed_request_count: int = 0
    total_input_tokens: int = 0
    total_output_tokens: int = 0
    total_cached_input_tokens: int = 0
    total_cost_nano_usd: int = 0
    p95_duration_ms: int | None = None
    cost_status_counts: dict[str, int] = Field(default_factory=dict)


class MonitoringToolOverview(BaseModel):
    tool_call_count: int = 0
    failed_call_count: int = 0
    denied_call_count: int = 0
    waiting_for_approval_count: int = 0
    approval_requested_count: int = 0
    approval_denied_count: int = 0
    p95_total_duration_ms: int | None = None
    p95_approval_wait_ms: int | None = None


class MonitoringModelSummary(BaseModel):
    provider_id: str
    model_id: str
    request_count: int
    retry_request_count: int
    total_cost_nano_usd: int


class MonitoringToolSummary(BaseModel):
    tool_name: str
    call_count: int
    failed_call_count: int
    denied_call_count: int
    average_total_duration_ms: int | None = None


class MonitoringOverviewResponse(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    project_id: str | None = None
    window_hours: int
    health: ObservabilityHealth
    llm: MonitoringLLMOverview
    tools: MonitoringToolOverview
    top_models: list[MonitoringModelSummary] = Field(default_factory=list)
    top_tools: list[MonitoringToolSummary] = Field(default_factory=list)


class MonitoringProviderRequestItem(BaseModel):
    id: str
    logical_call_id: str
    project_id: str | None = None
    session_id: str | None = None
    run_id: str | None = None
    provider_id: str | None = None
    model_id: str | None = None
    request_attempt_index: int
    status: str
    duration_ms: int | None = None
    total_cost_nano_usd: int | None = None
    cost_status: str
    finish_reason: str | None = None
    started_at: datetime
    error_message: str | None = None
    input_tokens: int | None = None
    output_tokens: int | None = None
    cached_input_tokens: int | None = None


class MonitoringProviderRequestListResponse(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    project_id: str | None = None
    window_hours: int
    total: int
    items: list[MonitoringProviderRequestItem]


class MonitoringToolCallItem(BaseModel):
    id: str
    invocation_id: str
    tool_call_id: str
    project_id: str | None = None
    session_id: str | None = None
    run_id: str | None = None
    tool_name: str
    status: str
    execution_duration_ms: int | None = None
    approval_wait_ms: int | None = None
    total_duration_ms: int | None = None
    terminal_reason: str | None = None
    error_category: str | None = None
    error_message: str | None = None
    latest_approval_event_type: str | None = None
    latest_approval_reason: str | None = None
    started_at: datetime
    finished_at: datetime | None = None


class MonitoringToolCallListResponse(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    project_id: str | None = None
    window_hours: int
    total: int
    items: list[MonitoringToolCallItem]


class MonitoringTrendPoint(BaseModel):
    bucket_start: datetime
    llm_request_count: int = 0
    llm_failed_count: int = 0
    llm_retry_count: int = 0
    llm_total_cost_nano_usd: int = 0
    tool_call_count: int = 0
    tool_failed_count: int = 0
    tool_denied_count: int = 0


class MonitoringTrendResponse(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    project_id: str | None = None
    window_hours: int
    bucket_hours: int
    points: list[MonitoringTrendPoint]


class MonitoringApprovalEventItem(BaseModel):
    id: str
    approval_id: str
    event_type: str
    actor_type: str | None = None
    reason: str | None = None
    occurred_at: datetime


class MonitoringProviderRequestDetail(MonitoringProviderRequestItem):
    pricing_id: str | None = None
    pricing_match_rule: str | None = None
    pricing_version: str | None = None
    input_price_nano_usd_per_million: int | None = None
    output_price_nano_usd_per_million: int | None = None
    cached_input_price_nano_usd_per_million: int | None = None
    input_cost_nano_usd: int | None = None
    output_cost_nano_usd: int | None = None
    cached_input_cost_nano_usd: int | None = None


class MonitoringToolCallDetail(MonitoringToolCallItem):
    approval_events: list[MonitoringApprovalEventItem] = Field(default_factory=list)


class MonitoringModelAnomaly(BaseModel):
    provider_id: str
    model_id: str
    request_count: int
    retry_request_count: int
    failed_request_count: int
    incomplete_cost_count: int
    total_cost_nano_usd: int


class MonitoringToolAnomaly(BaseModel):
    tool_name: str
    call_count: int
    failed_call_count: int
    denied_call_count: int
    waiting_for_approval_count: int
    average_approval_wait_ms: int | None = None


class MonitoringAnomalyResponse(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    project_id: str | None = None
    window_hours: int
    incomplete_cost_request_count: int = 0
    interrupted_request_count: int = 0
    waiting_approval_call_count: int = 0
    hottest_retry_models: list[MonitoringModelAnomaly] = Field(default_factory=list)
    noisiest_tools: list[MonitoringToolAnomaly] = Field(default_factory=list)


class MonitoringAlertState(BaseModel):
    key: str
    severity: str
    title: str
    current_value: int
    threshold_value: int
    description: str


class MonitoringAlertStatusResponse(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    settings: MonitoringAlertSettings
    active_alerts: list[MonitoringAlertState] = Field(default_factory=list)

File: frontend/src/types/monitoring.ts
export type MonitoringHealthStatus = 'healthy' | 'degraded' | 'critical'

export interface MonitoringHealth {
  status: MonitoringHealthStatus
  last_event_recorded_at?: string | null
  last_projection_at?: string | null
  projection_lag_count: number
  fallback_backlog_count: number
  memory_queue_depth: number
  dropped_metrics_count: number
  last_error_code?: string | null
  last_error_at?: string | null
}

export interface MonitoringLLMOverview {
  logical_call_count: number
  provider_request_count: number
  retry_request_count: number
  failed_request_count: number
  total_input_tokens: number
  total_output_tokens: number
  total_cached_input_tokens: number
  total_cost_nano_usd: number
  p95_duration_ms?: number | null
  cost_status_counts: Record<string, number>
}

export interface MonitoringToolOverview {
  tool_call_count: number
  failed_call_count: number
  denied_call_count: number
  waiting_for_approval_count: number
  approval_requested_count: number
  approval_denied_count: number
  p95_total_duration_ms?: number | null
  p95_approval_wait_ms?: number | null
}

export interface MonitoringModelSummary {
  provider_id: string
  model_id: string
  request_count: number
  retry_request_count: number
  total_cost_nano_usd: number
}

export interface MonitoringToolSummary {
  tool_name: string
  call_count: number
  failed_call_count: number
  denied_call_count: number
  average_total_duration_ms?: number | null
}

export interface MonitoringOverviewResponse {
  project_id?: string | null
  window_hours: number
  health: MonitoringHealth
  llm: MonitoringLLMOverview
  tools: MonitoringToolOverview
  top_models: MonitoringModelSummary[]
  top_tools: MonitoringToolSummary[]
}

export interface MonitoringProviderRequestItem {
  id: string
  logical_call_id: string
  project_id?: string | null
  session_id?: string | null
  run_id?: string | null
  provider_id?: string | null
  model_id?: string | null
  request_attempt_index: number
  status: string
  duration_ms?: number | null
  total_cost_nano_usd?: number | null
  cost_status: string
  finish_reason?: string | null
  started_at: string
  error_message?: string | null
  input_tokens?: number | null
  output_tokens?: number | null
  cached_input_tokens?: number | null
}

export interface MonitoringProviderRequestListResponse {
  project_id?: string | null
  window_hours: number
  total: number
  items: MonitoringProviderRequestItem[]
}

export interface MonitoringToolCallItem {
  id: string
  invocation_id: string
  tool_call_id: string
  project_id?: string | null
  session_id?: string | null
  run_id?: string | null
  tool_name: string
  status: string
  execution_duration_ms?: number | null
  approval_wait_ms?: number | null
  total_duration_ms?: number | null
  terminal_reason?: string | null
  error_category?: string | null
  error_message?: string | null
  latest_approval_event_type?: string | null
  latest_approval_reason?: string | null
  started_at: string
  finished_at?: string | null
}

export interface MonitoringToolCallListResponse {
  project_id?: string | null
  window_hours: number
  total: number
  items: MonitoringToolCallItem[]
}

export interface MonitoringTrendPoint {
  bucket_start: string
  llm_request_count: number
  llm_failed_count: number
  llm_retry_count: number
  llm_total_cost_nano_usd: number
  tool_call_count: number
  tool_failed_count: number
  tool_denied_count: number
}

export interface MonitoringTrendResponse {
  project_id?: string | null
  window_hours: number
  bucket_hours: number
  points: MonitoringTrendPoint[]
}

export interface MonitoringApprovalEventItem {
  id: string
  approval_id: string
  event_type: string
  actor_type?: string | null
  reason?: string | null
  occurred_at: string
}

export interface MonitoringProviderRequestDetail extends MonitoringProviderRequestItem {
  pricing_id?: string | null
  pricing_match_rule?: string | null
  pricing_version?: string | null
  input_price_nano_usd_per_million?: number | null
  output_price_nano_usd_per_million?: number | null
  cached_input_price_nano_usd_per_million?: number | null
  input_cost_nano_usd?: number | null
  output_cost_nano_usd?: number | null
  cached_input_cost_nano_usd?: number | null
}

export interface MonitoringToolCallDetail extends MonitoringToolCallItem {
  approval_events: MonitoringApprovalEventItem[]
}

export interface MonitoringModelAnomaly {
  provider_id: string
  model_id: string
  request_count: number
  retry_request_count: number
  failed_request_count: number
  incomplete_cost_count: number
  total_cost_nano_usd: number
}

export interface MonitoringToolAnomaly {
  tool_name: string
  call_count: number
  failed_call_count: number
  denied_call_count: number
  waiting_for_approval_count: number
  average_approval_wait_ms?: number | null
}

export interface MonitoringAnomalyResponse {
  project_id?: string | null
  window_hours: number
  incomplete_cost_request_count: number
  interrupted_request_count: number
  waiting_approval_call_count: number
  hottest_retry_models: MonitoringModelAnomaly[]
  noisiest_tools: MonitoringToolAnomaly[]
}

export interface MonitoringAlertSettings {
  enable_in_app_notifications: boolean
  poll_interval_seconds: number
  enable_webhook_notifications: boolean
  webhook_url?: string | null
  webhook_min_severity: string
  webhook_cooldown_seconds: number
  retry_request_count_warn: number
  failed_request_count_warn: number
  incomplete_cost_request_count_warn: number
  tool_failed_call_count_warn: number
  approval_denied_count_warn: number
  approval_wait_p95_ms_warn: number
  projection_lag_count_warn: number
  fallback_backlog_count_warn: number
  memory_queue_depth_critical: number
}

export interface MonitoringAlertState {
  key: string
  severity: string
  title: string
  current_value: number
  threshold_value: number
  description: string
}

export interface MonitoringAlertStatusResponse {
  settings: MonitoringAlertSettings
  active_alerts: MonitoringAlertState[]
}
